### PR TITLE
Barplot viz

### PR DESCRIPTION
**Pull request by moontrip**
With the addition of a barplot component at web-components, basic barplot viz is made. Certainly, this PR requires the approve/merge the [barplot component PR](https://github.com/VEuPathDB/web-components/pull/124).

A screenshot is attached here:

barplot1

### CODE REVIEW

**Comment by moontrip** [comment]
@bobular @dmfalke Somehow slack is down so I just pushed a new commit here with the change of default barLayout from overlay to group. With the same variables shown by Bob, here is the screenshot with group:

barplot-overlaygroup

**Comment by moontrip** [comment]
I could also confirm that the latest change by Danielle (or someone else) to add `proportion` option (for config.valueSpec prop in Post), in conjunction with `count`, works - shows from 0 to 1 in y-axis with the `proportion` option

**Comment by moontrip** [comment]
a new commit is made for adopting variable constraints

**Comment by moontrip** [comment]
empty data (or data error) with empty plot is implemented. Screenshots of full screen and grid views are attached:

barplot-empty1
barplot-empty

File: src/lib/core/api/data-api.ts
Comment (bobular) [inline src/lib/core/api/data-api.ts:164]: The [API Spec](https://veupathdb.github.io/EdaDataService/api.html#resources:/apps/pass/visualizations/barplot:post) says label is an array of string.

What discussions have you had with @d-callan about this?  I see that the labels of variables such as `GEMS->Participant->Enrollment year` are being returned as numbers.

Should the API specs be changed or should the service return strings?

I'm just making sure everyone has had a chance to discuss this.

File: src/lib/core/components/filter/TableFilter.tsx
Comment (bobular) [inline src/lib/core/components/filter/TableFilter.tsx:99]: See the comment on data-api.ts - **if** we roll that back, don't forget to roll this back.

**Review by bobular — CHANGES_REQUESTED** [review]
Looks great DK.
See the questions about validation of `label: array(string)`

Also, in FF and Chrome, I see the page footer hidden underneath the plot.
Please also tweak the thumbnail plot size for grid view - it looks to me a bit off-centre and could fill up the box more?

File: src/lib/core/api/data-api.ts
Comment (moontrip) [reply src/lib/core/api/data-api.ts:164]: @bobular, first of all, thank you for your review 👍 

As for this issue, I know that the API doc shows the Label as string/string[], however, from my limited tests, it always returned number/number[]. I did not specifically ask Danielle about this as there have been many discrepancies between the doc and the return. Thus, I used union type instead, but we should ask Danielle about this as well.

File: src/lib/core/components/filter/TableFilter.tsx
Comment (moontrip) [reply src/lib/core/components/filter/TableFilter.tsx:99]: @bobular as you said, this is related to above. If we can confirm with Danielle whether the label is either string[] or number[], then I can use single type for label in there (and at data-api.ts).

**Comment by moontrip** [comment]
@bobular 

> Also, in FF and Chrome, I see the page footer hidden underneath the plot.

**DK:** I am sorry that I cannot understand the issue. The page footer you meant is probably the bottom banner with VEuPathDB icons, but it shows me correctly for both FF and Chrome. Here is a screenshot at FF (I have reduced the browser size so that the captured image is not that big). Please correct me if I misunderstood your point.

barplot-ff2

> Please also tweak the thumbnail plot size for grid view - it looks to me a bit off-centre and could fill up the box more?

**DK:** I got your point. It is because I have set margin based on empty plot with X/Y ticklabels: you have changed them without ticklabels: will adjust it. 
By the way, since we are subject to thumbnail/gridview, I want to ask you about one thing. For barplot, there are many cases that X tick labels are quite lengthy (e.g., see my screenshot for X - breastfead & overlay - country). So I have been thinking that I should hide (at least) X tick labels for barplot's grid view. If that is the case, I need to introduce additional prop to show/hide axis tick label. What do you think?

barplot-gridview1

**Comment by bobular** [comment]
Hi DK.
My replies in reverse order...

Yes, good idea to turn off X (and yes, I guess Y too) labels in thumbnail view.

Actually, after my changes, the empty plots do have ticklabels - but the font color is transparent. So it shouldn't affect spacing and layout?

Regarding the footer layout issue, here's a screenshot from my Linux FF:

image

It seems that if I had a taller screen, the problem would go away.  The footer appears at the bottom of the browser window, and then the barplot is drawn over the top of it.  Histogram and Scatter don't do this. (Haven't checked Mosaic!)

**Comment by moontrip** [comment]
Thank you for your response, @bobular. Actually I cannot reproduce your case at my FF and Chrome (both at linux/Ubuntu VB). I tried to make very small height of browser, but it always show browser's scrollbar to separate (see my screenshot below). 
I will also add show/hide tick label props for both indep. and dep. axis.

barplot-scroll

**Comment by dmfalke** [comment]
We should definitely clarify the API with Danielle. I haven’t looked at the
footer issue yet. Otherwise this is looking good!

**Comment by moontrip** [comment]
@dmfalke Thank you for your review 👍 

By the way, may I ask one thing? If we have multiple reviewers (like this PR having three three reviewers), does a merge need all approvals from all reviewers or it can be made by one approval (from one reviewer) ? Actually similar situation happens on my PR regarding [barplot component PR](https://github.com/VEuPathDB/web-components/pull/124) at web-component repo (here): Bob approved the PR but not quite sure if I need both you and Jamie's approvals for merge as well.

**Comment by dmfalke** [comment]
I think that should be the policy. Sorry for the confusion and delay.

**Comment by bobular** [comment]
One thing (by email) - is the big Barplot header still there?  The other
plots don't have it.  Could you remove it please?

**Comment by moontrip** [comment]
Good evening, Bob

I missed that! Thank you for finding. Will remove it.

**Comment by moontrip** [comment]
Update:
a) implemented to hide axis tick labels for grid view/thumbnail (through [additional props](https://github.com/VEuPathDB/web-eda/pull/92/commits/acd6d37cefd635782693b52f558f57a9ab0416ab) at bar plot component); b) adjusted plot size and margin for grid view/thumbnail; c) removed big "Bar plot" header at full screen view (reported by Bob)

**Review by dmfalke — APPROVED** [review]
I'm okay with this being merged, as long as we make an issue in this repo to update the response type once we get clarity the intended type. Waiting is also fine. Whichever is more comfortable for you, @moontrip. Thanks!

**Comment by moontrip** [comment]
> I'm okay with this being merged, as long as we make an issue in this repo to update the response type once we get clarity the intended type. Waiting is also fine. Whichever is more comfortable for you, @moontrip. Thanks!

Thank you for your review and comment @dmfalke 👍 Since the type change can be easily addressed later after merge and I don't know when Danielle will change it, it may be better to merge this and be introduced at feature site for others to test it, potentially revealing/improving something else we have missed.

**Review by jtlong3rd — APPROVED** [review]

**Review by bobular — APPROVED** [review]

**Pull request merged**
